### ISSUE TITLE
Some issues with FitTrackee nginx deployment (logging and file upload limit size)

### ISSUE THREAD

**Issue by jat255** [SamR1/FitTrackee#256]
This issue is more of a debugging help, and is in a few different parts that I'm trying to figure out. Apologies if it's hard to follow, but I think the issue is in my deployment and I can't see where the underlying problem is. I've broken it up into a summary of my debugging process thus far.

## Part 1: Cannot upload large GPX file even after changing limit

I've run into some trouble developing my Strava-to-FitTrackee adapter, in that I'm unable to upload GPX files larger than 1 MB, even though I've set the limit higher in my web UI:

![image](https://user-images.githubusercontent.com/1278301/199762128-fe1e1b8b-2868-4b59-a259-6c39775d523f.png)

I've confirmed in the backend database that those values were saved:

```
fittrackee=> SELECT * FROM app_config;
 id | max_users | gpx_limit_import | max_single_file_size | max_zip_file_size | admin_contact 
----+-----------+------------------+----------------------+-------------------+---------------
  1 |         0 |              100 |             20971520 |         104857600 | 
(1 row)
```

However, when I go to upload a GPX using the API, I receive a `413 Request Entity Too Large` (`test_file.gpx` is 1,145,739 bytes (1.1 MB)):

```
$ curl --request POST \
  --url https://fittrackee.host/api/workouts \
  --header 'Authorization: Bearer XXXXXXXXXXXXXXXXXXXXXXXXXXX' \
  --header 'Content-Type: multipart/form-data' \
  --form file=@test_file.gpx \
  --form 'data={"sport_id": 1}'
<html>
<head><title>413 Request Entity Too Large</title></head>
<body>
<center><h1>413 Request Entity Too Large</h1></center>
<hr><center>nginx/1.18.0</center>
</body>
</html>
```

## Part 2: Logs are not working as I'd expect

I have my deployment set up via systemd as described in the installation instructions. I installed FitTrackee on this server via `pip` using a virtualenv present at `/home/fittrackee/.venv`. Here are my unit files (sanitized) and nginx config:

<details>
  <summary><code>/etc/systemd/system/fittrackee.service</code>:</summary>
  
  ```
  [Unit]
  Description=FitTrackee service
  After=network.target
  After=postgresql.service
  # After=redis.service
  StartLimitIntervalSec=0
  
  [Service]
  Type=simple
  Restart=always
  RestartSec=1
  User=fittrackee
  StandardOutput=journal
  StandardError=journal
  SyslogIdentifier=fittrackee
  Environment="APP_SECRET_KEY=xxxxxxxxxxx"
  Environment="APP_LOG=/home/fittrackee/fittrackee.log"
  Environment="UPLOAD_FOLDER=/mnt/fittrackee/uploads"
  Environment="DATABASE_URL=postgresql://xxxx:xxxxxxxxxx@127.0.0.1:5432/fittrackee"
  Environment="UI_URL="
  Environment="EMAIL_URL="
  Environment="SENDER_EMAIL="
  Environment="DEFAULT_STATICMAP=True"
  #Environment="REDIS_URL="
  #Environment="TILE_SERVER_URL="
  Environment="STATICMAP_SUBDOMAINS="
  Environment="MAP_ATTRIBUTION="
  Environment="WEATHER_API_KEY="
  WorkingDirectory=/home/fittrackee
  ExecStart=/home/fittrackee/.venv/bin/gunicorn -b 127.0.0.1:5000 "fittrackee:create_app()" --error-logfile /home/fittrackee/gunicorn.log
  Restart=always
  
  [Install]
  WantedBy=multi-user.target
  ```
</details>

<details>
  <summary><code>/etc/systemd/system/fittrackee_workers.service</code>:</summary>
  
  ```
  [Unit]
  Description=FitTrackee task queue service
  After=network.target
  After=postgresql.service
  # After=redis.service
  StartLimitIntervalSec=0
  
  [Service]
  Type=simple
  Restart=always
  RestartSec=1
  User=fittrackee
  StandardOutput=journal
  StandardError=journal
  SyslogIdentifier=fittrackee_workers
  Environment="FLASK_APP=fittrackee"
  Environment="APP_SECRET_KEY=xxxxxxxxxx"
  Environment="APP_LOG=/home/fittrackee/fittrackee_workers.log"
  Environment="UPLOAD_FOLDER=/mnt/fittrackee/uploads"
  Environment="DATABASE_URL=postgresql://xxxxxxx:xxxxxxxxxxx@127.0.0.1:5432/fittrackee"
  Environment="UI_URL="
  Environment="EMAIL_URL="
  Environment="SENDER_EMAIL="
  # Environment="REDIS_URL="
  WorkingDirectory=/home/fittrackee
  ExecStart=/home/fittrackee/.venv/bin/flask worker --processes 2
  Restart=always
  
  [Install]
  WantedBy=multi-user.target
  ```
</details>

<details>
  <summary><code>/etc/nginx/sites-enabled/fittrackee</code>:</summary>
  
  ```
  server {
    listen 443 ssl http2;
    server_name <my_ip>;
    ssl_certificate /home/fittrackee/fullchain.pem;
    ssl_certificate_key /home/fittrackee/privkey.pem;

    location / {
        proxy_pass http://127.0.0.1:5000;
        proxy_redirect    default;
        proxy_set_header  Host $host;
        proxy_set_header  X-Real-IP $remote_addr;
        proxy_set_header  X-Forwarded-For $proxy_add_x_forwarded_for;
        proxy_set_header  X-Forwarded-Host $server_name;
        proxy_set_header  X-Forwarded-Proto $scheme;
    }
}

server {
    listen 80;
    server_name <my_ip>;
    location / {
        return 301 https://<my_ip>$request_uri;
    }
}
  ```
</details>

To start tracking things down, I started placing various `appLog.warning()` calls in files within `/home/fittrackee/.venv/lib/python3.9/site-packages/fittrackee`, when I noticed something interesting. Logging statements placed at the top level of modules would log to `fittrackee.log` as expected when the modules were imported, but any logging done within methods (such as the `post_workout()` method) would not show up in the log (even after restarting the systemd service). 

## Part 3: It all works if I run `gunicorn` directly from the command line, or remove `nginx` from the equation

To simplify things, I tried doing away with systemd and nginx, and ran the app directly from the command line with the following:

```
$ APP_SECRET_KEY=xxxxxxxxx APP_LOG=/home/fittrackee/fittrackee.log UPLOAD_FOLDER=/mnt/fittrackee/uploads DATABASE_URL=postgresql://xxxxxxxx:xxxxxxxxxxx@127.0.0.1:5432/fittrackee DEFAULT_STATICMAP=True /home/fittrackee/.venv/bin/gunicorn -b <local_ip_address>:5000 "fittrackee:create_app()"  --error-logfile -
```

To my surprise, everything worked as expected. I could upload my GPX file using the `<local_ip_address>:5000/api/workouts` endpoint (the limit was properly set), and my `appLog.warning()` statements inside the various methods were logged as expected to `/home/fittrackee/fittrackee.log`.

Likewise, if I replace the `ExecStart=` line in my systemd unit file with `/home/fittrackee/.venv/bin/gunicorn -b <local_ip_address>:5000 "fittrackee:create_app()"  --error-logfile -`, it also all appears to work as expected when I access the app via the local IP and port.

## Conclusion

At this point, I'm at a loss of why using nginx is causing funky behavior. It's almost as if it's caching the app, I don't think that's the case, as some log statements (those at the top of files) will work, and others don't (those within methods). I don't understand why the `max_single_file_size` is not being respected when routing things through nginx.

Sorry for the long-winded post... any ideas?

**Comment by SamR1** [comment]
Hi,

> However, when I go to upload a GPX using the API, I receive a `413 Request Entity Too Large` (`test_file.gpx` is 1,145,739 bytes (1.1 MB)):
> 
> ```
> $ curl --request POST \
>   --url https://fittrackee.host/api/workouts \
>   --header 'Authorization: Bearer XXXXXXXXXXXXXXXXXXXXXXXXXXX' \
>   --header 'Content-Type: multipart/form-data' \
>   --form file=@test_file.gpx \
>   --form 'data={"sport_id": 1}'
> <html>
> <head><title>413 Request Entity Too Large</title></head>
> <body>
> <center><h1>413 Request Entity Too Large</h1></center>
> <hr><center>nginx/1.18.0</center>
> </body>
> </html>
> ```

It looks like nginx is returning the error, not the API.
When it's the API, the response looks like: 
```
{"status": "fail", "message": "Error during workout upload, file size (12.9MB) exceeds 10.0MB."}
```

Can you check [`client_max_body_size`](https://nginx.org/en/docs/http/ngx_http_core_module.html#client_max_body_size) in nginx configuration? By default it's set to `1m`.

**Comment by jat255** [comment]
> Can you check [`client_max_body_size`](https://nginx.org/en/docs/http/ngx_http_core_module.html#client_max_body_size) in nginx configuration?

Doh! Of course that was it. I set `client_max_body_size 0;` in my `server` config block and all's working correctly now. This might be worth a note in the documentation as potential "gotcha"

**Comment by jat255** [comment]
> This might be worth a note in the documentation as potential "gotcha"

And now that I went in to add a note, I see there's one there already and I just need to read better... Thanks again. I've submitted a PR to make the note a little more explicit in the Nginx config example. I think what happened was I just copy/pasted that config and then didn't read (or forgot about) the note afterwards.
